overlord/devicestate/devicestate.go

```diff
@@ -1689,6 +1684,19 @@ func CreateRecoverySystem(st *state.State, label string, opts CreateRecoverySyst
 			continue
 		}
 
+		if sn.Presence != "required" {
+			sets, _, err := valsets.CheckPresenceRequired(sn)
```

Comment:
not too familiar with validation sets, but can the snap have an 'invalid' presence?

Reply:
Yes, they can. If any snaps in the model are marked invalid in any validation sets, that is caught above in `checkForInvalidSnapsInModel`.

Reply:
yes, but sn is a ModelSnap  and in the model snaps can only be marked as either "required" or "optional"